meadow-client/src/Blockly/Generator.purs

```diff
@@ -0,0 +1,119 @@
+module Blockly.Generator where
+
+import Prelude
+import Blockly.Types (Block, BlocklyState, Workspace)
+import Data.Array as Array
+import Data.Either (Either(..))
+import Data.Function.Uncurried (Fn1, Fn4, Fn5, Fn6, runFn1, runFn4, runFn5, runFn6)
+import Data.Maybe (Maybe)
+import Effect (Effect)
+import Effect.Uncurried (EffectFn1, EffectFn2, EffectFn4, runEffectFn1, runEffectFn2, runEffectFn4)
+import Partial.Unsafe (unsafePartial)
+
+type GeneratorFunction
+  = BlocklyState -> Block -> Either String String
+
+data Order
+  = Atomic
+  | None
+
+toNumber :: Order -> Number
+toNumber Atomic = 0.0
+
+toNumber None = 99.0
+
+foreign import data Generator :: Type
+
+foreign import data Input :: Type
+
+foreign import data Field :: Type
+
+foreign import getFieldValue_ :: forall a. Fn4 (String -> Either String a) (a -> Either String a) Block String (Either String String)
+
+foreign import statementToCode_ :: forall a. Fn5 (String -> Either String a) (a -> Either String a) Generator Block String (Either String String)
+
+foreign import valueToCode_ :: forall a. Fn6 (String -> Either String a) (a -> Either String a) Generator Block String Number (Either String String)
+
+foreign import mkGenerator_ :: EffectFn2 BlocklyState String Generator
+
+foreign import insertGeneratorFunction_ :: EffectFn4 BlocklyState Generator String (BlocklyState -> Block -> String) Unit
+
+foreign import workspaceToCode_ :: EffectFn2 BlocklyState Generator String
+
+foreign import inputList_ :: Fn1 Block (Array Input)
+
+foreign import connectToPrevious_ :: EffectFn2 Block Input Unit
+
+foreign import connectToOutput_ :: EffectFn2 Block Input Unit
+
+foreign import newBlock_ :: EffectFn2 Workspace String Block
+
+foreign import inputName_ :: Fn1 Input String
+
+foreign import clearWorkspace_ :: EffectFn1 Workspace Unit
+
+foreign import fieldRow_ :: Fn1 Input (Array Field)
+
+foreign import setFieldText_ :: EffectFn2 Field String Unit
+
+foreign import fieldName_ :: Fn1 Field String
+
+foreign import unsafeThrowError_ :: forall a. Fn1 String a
+
+getFieldValue :: Block -> String -> Either String String
+getFieldValue = runFn4 getFieldValue_ Left Right
+
+statementToCode :: Generator -> Block -> String -> Either String String
+statementToCode = runFn5 statementToCode_ Left Right
+
+valueToCode :: Generator -> Block -> String -> Order -> Either String String
+valueToCode g b v o = runFn6 valueToCode_ Left Right g b v (toNumber o)
+
+mkGenerator :: BlocklyState -> String -> Effect Generator
+mkGenerator = runEffectFn2 mkGenerator_
+
+insertGeneratorFunction :: BlocklyState -> Generator -> String -> GeneratorFunction -> Effect Unit
+insertGeneratorFunction bs g k f = runEffectFn4 insertGeneratorFunction_ bs g k (compose2 (unsafePartial rightOrDie) f)
+
+rightOrDie :: forall a. Partial => Either String a -> a
+rightOrDie (Right a) = a
```

Comment:
`unsafeFromRight` would seem more conventional.

Reply:
oooh, didn't know about that.

Reply:
oh, just the name :(

Reply:
Oh look, here it is:

https://pursuit.purescript.org/packages/purescript-either/4.1.1/docs/Data.Either#v:fromRight

I didn't notice because they drop the `unsafe`, relying on the `Partial` instead.

Reply:
it's not quite the same, I want to throw the `Left` value in the error so I can see exactly what went wrong in the console